### PR TITLE
fixes #22467 - dashboard: list hosts with expired token

### PR DESCRIPTION
**Pull request by timogoebel**
image
image

### CODE REVIEW

File: app/views/dashboard/_hosts_expired_token_widget_host_list.html.erb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    <th><%= _('Owner') %></th>
+    <th><%= _('Created') %></th>
+    <th><%= _('Token Expired') %></th>
+  </thead>
+  <tbody>
+    <% hosts.each do |host| %>
+      <tr>
+        <td class="ellipsis"><%= icon_text('error-circle-o', '', kind: 'pficon') %> <%= link_to host.name, host_path(host) %></td>
+        <td class="hidden-tablet hidden-xs ellipsis"><%= host.owner %></td>
+        <td class="hidden-tablet hidden-xs ellipsis"><%= date_time_relative_value(host.created_at) %></td>
```
Comment (sean797) [inline app/views/dashboard/_hosts_expired_token_widget_host_list.html.erb]: Would getting a value for "Built" be better then created_at? Just created_at wouldn't be relevant to hosts that have been rebuilt.

You could use Setting[:token_duration] and `host.token.expires` to work it out. Obviously that would be incorrect if someone changes Setting[:token_duration] though :(

**Comment by ShimShtein** [comment]
Nice! Can we also add a scoped search definition for `expired?`?

**Review by dLobatog — DISMISSED** [review]
Setting to WoC based on comments

File: app/views/dashboard/_hosts_expired_token_widget_host_list.html.erb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<table class="table table-striped table-fixed">
```
Comment (ohadlevy) [inline app/views/dashboard/_hosts_expired_token_widget_host_list.html.erb]: should we use the table css classes helper here?

**Comment by timogoebel** [comment]
I changed the code to use the helper and removed the created at column in the table for now. I think, we should store the time a host is set to built in the database. But I want to do that in a separate PR. I'll also create an issue to add `expired?` to scoped search. This is out of scope here (although a good idea).

Anything else to add?

This is how it looks now:
image

**Comment by timogoebel** [comment]
@lzap actually added the required fields to show the build start time in #5292. I'll wait for the PR to get in before I continue to work on this.

**Comment by lzap** [comment]
Nice patch, I'd also like to see list of hosts which are actually in build mode. Do you mind renaming the feature and widget to show both - there could be a filter or button to only show expired ones. I don't know how others, but my deployments have usually just few hosts in build mode.

**Comment by timogoebel** [comment]
> I'd also like to see list of hosts which are actually in build mode.

Nice idea, I'll look into it.

**Comment by timogoebel** [comment]
Updated. This is how it currently looks.
@lzap: Do you mind taking a look? Should be an easy review. :-)

image

**Comment by ohadlevy** [comment]
@timogoebel last screenshot is very nice :) would you consider having its own page for it (vs a dashboard widget?)

**Comment by timogoebel** [comment]
> would you consider having its own page for it (vs a dashboard widget?)

I prefer the widget for now as it's more prominent (and because I don't have time for more), but we could go crazy on a separate page (e.g. with some visual indication of what is actually happening, something like this: https://res.infoq.com/news/2016/02/netflix-spinnaker/en/resources/1spinnaker-pipelines.png). I'd really like this but I believe that's a huge amount of work to accomplish.

File: app/views/dashboard/_hosts_in_build_mode_widget_host_list.html.erb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  </thead>
+  <tbody>
+    <% hosts.each do |host| %>
+      <tr>
+        <td class="ellipsis"><%= host_build_status_icon(host) %> <%= link_to host.name, host_path(host) %></td>
+        <td class="hidden-tablet hidden-xs ellipsis"><%= host.owner %></td>
+        <td class="hidden-tablet hidden-xs ellipsis"><%= build_duration(host) %></td>
+        <td class="hidden-tablet hidden-xs ellipsis"><%= date_time_relative_value(host.token.expires) %></td>
+      </tr>
+  <% end %>
```
Comment (ekohl) [inline app/views/dashboard/_hosts_in_build_mode_widget_host_list.html.erb]: nit: odd indenting

File: app/helpers/dashboard_helper.rb [hunk 4]
```diff
@@ -159,6 +159,20 @@ def out_of_sync_enabled?(origin)
...
+      label = _('Token expired')
+    elsif host.build_errors.present?
+      icon = 'error-circle-o'
+      label = _('Build errors')
+    else
+      icon = 'in-progress'
+      label = _('Build in progress')
+    end
+    icon_text(icon, '', kind: 'pficon', title: label)
+  end
```
Comment (lzap) [inline app/helpers/dashboard_helper.rb]: Perhaps to use different icon for expired and errored states? Just a thought.

**Review by lzap — DISMISSED** [review]
Nice, please rebase the indenting and I am gonna test and merge thanks.

File: app/helpers/dashboard_helper.rb (hunk 4 above)
Comment (timogoebel) [reply app/helpers/dashboard_helper.rb]: Patternfly doesn't have a icon that fits in my opinion.
https://www.patternfly.org/styles/icons/
Anything you like, @lzap?

**Comment by timogoebel** [comment]
Fixed the indent issue.

File: app/helpers/dashboard_helper.rb (hunk 4 above)
Comment (ohadlevy) [reply app/helpers/dashboard_helper.rb]: @rohoover any suggestions?

File: app/helpers/dashboard_helper.rb [hunk 5]
```diff
@@ -159,6 +159,20 @@ def out_of_sync_enabled?(origin)
     setting.nil? ? true : !setting
   end
 
+  def host_build_status_icon(host)
+    if host.token_expired?
+      icon = 'error-circle-o'
+      label = _('Token expired')
+    elsif host.build_errors.present?
+      icon = 'error-circle-o'
+      label = _('Build errors')
```
Comment (lzap) [inline app/helpers/dashboard_helper.rb]: There is one problem here - if a host has `build_errors` set, it is never in `build` mode. When I error a host by running this command:

    wget -q -O- --body-file=LICENSE --method POST --header="Content-Type: text/plain" --no-check-certificate http://localhost:3000/unattended/failed?token=689ba68a-a511-44b1-bca5-c8f3f1ff353e

then the host disappears. Therefore this branch can't be reached.

File: app/helpers/dashboard_helper.rb (hunk 4 above)
Comment (lzap) [reply app/helpers/dashboard_helper.rb]: How about the bell icon? Like expired = bell = ring = end. "fa-bell" is the name.

File: app/views/dashboard/_hosts_in_build_mode_widget.html.erb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<% hosts = @data.hosts.in_build_mode.includes(:token).preload(:owner).order(created_at: :desc).limit(9) %>
```
Comment (lzap) [inline app/views/dashboard/_hosts_in_build_mode_widget.html.erb]: This query must include both "in build" and also when "build errors" text is present (not empty).

File: app/views/dashboard/_hosts_in_build_mode_widget.html.erb [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+<h4 class="header">
+  <%= _('Hosts in Build mode') %>
```
Comment (lzap) [inline app/views/dashboard/_hosts_in_build_mode_widget.html.erb]: This could be also "Hosts in Build mode or Failed hosts" or similar.

**Review by lzap — CHANGES_REQUESTED** [review]
One remark.

File: app/helpers/dashboard_helper.rb (hunk 4 above)
Comment (Rohoover) [reply app/helpers/dashboard_helper.rb]: @lzap 

We use the bell icon for the notification drawer. I wouldn't repeat it. I'm bringing this back to the team to see if it's already been identified elsewhere. Stay tuned!

Comment (Rohoover) [reply app/helpers/dashboard_helper.rb]: This is what I'm proposing:

Expiring soon = fa-hourglass-end
Expired = fa-hourglass-o

File: app/helpers/dashboard_helper.rb (hunk 5 above)
Comment (timogoebel) [reply app/helpers/dashboard_helper.rb]: Interestingly, I get this:

```
Encoding::UndefinedConversionError ("\xE2" from ASCII-8BIT to UTF-8):
```

Maybe we need something like this:

```ruby
.encode('utf-8', invalid: :replace, undef: :replace, replace: '_')
```

`.gitignore` works fine, our `LICENSE` does not...

File: app/helpers/dashboard_helper.rb (hunk 4 above)
Comment (timogoebel) [reply app/helpers/dashboard_helper.rb]: Thanks, @Rohoover good call.

**Comment by timogoebel** [comment]
Thanks, added hosts with build errors to the widget and fixed an issue where hosts that entered build mode before this patch was applied showed an error. Also changed the icon to follox Rox' recommendation.

This is the latest screenshot:
image

**Review by lzap — APPROVED** [review]

**Comment by lzap** [comment]
Thank you for your contribution, merged.

File: app/helpers/hosts_helper.rb [hunk 8]
```diff
@@ -251,7 +251,7 @@ def name_field(host)
   end
 
   def build_duration(host)
-    return _("N/A") if host.initiated_at.nil? && host.installed_at.nil?
+    return _('N/A') if host.initiated_at.nil? || host.installed_at.nil?
     if host.installed_at.nil?
       time_ago_in_words(host.initiated_at, include_seconds: true) + " (in progress)"
```
Comment (alexjfisher) [inline app/helpers/hosts_helper.rb:256]: This code is now unreachable?

Comment (timogoebel) [reply app/helpers/hosts_helper.rb:256]: Looks unreachable, yes.

Comment (lzap) [reply app/helpers/hosts_helper.rb:256]: Bite me, but I do see a path: when `host.installed_at.nil?` is true?!? Explain please.

Comment (alexjfisher) [reply app/helpers/hosts_helper.rb:256]: @lzap Function will return on line 254 if host.installed_at.nil? is true.

Comment (lzap) [reply app/helpers/hosts_helper.rb:256]: I see.

Comment (lzap) [reply app/helpers/hosts_helper.rb:256]: I will take a look next week. My GH backlog is not getting slimmer.

Comment (lzap) [reply app/helpers/hosts_helper.rb:256]: Solved in https://github.com/theforeman/foreman/pull/6603

**Pull request merged**